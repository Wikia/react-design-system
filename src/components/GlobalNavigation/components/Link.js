import React from 'react';

import linkType from '../linkType';

/* eslint-disable react/prop-types */

function renderLinkNav({href, title, tracking_label}, props) {
  return (
    <a
      className="wds-global-navigation__link"
      href={href}
      data-tracking-label={tracking_label}
      {...props}
    >
      {title}
    </a>
  );
}

function renderLinkText({href, title, tracking_label}, props) {
  return (
    <a
      href={href}
      data-tracking-label={tracking_label}
      {...props}
    >
      {title}
    </a>
  );
}

function renderLinkButton({href, title, tracking_label}, props) {
  return (
    <a
      className="wds-global-navigation__link-button wds-button wds-is-secondary"
      href={href}
      data-tracking-label={tracking_label}
      {...props}
    >
      {title}
    </a>
  );
}

function renderLinkGroup({title, items}, props) {
  return (
    <div className="wds-global-navigation__link-group wds-has-shadow wds-dropdown">
      <div className="wds-global-navigation__dropdown-toggle wds-global-navigation__link wds-dropdown__toggle">
        <span>{title}</span>
        <svg className="wds-icon wds-icon-tiny wds-dropdown__toggle-chevron">
          <use xlinkHref="#wds-icons-dropdown-tiny" />
        </svg>
      </div>
      <div className="wds-global-navigation__dropdown-content wds-dropdown__content">
        <ul className="wds-list wds-is-linked">
          {items.map(item => (
            <li key={item.tracking_label}>
              <Link link={item} />
            </li>
          ))}
        </ul>
      </div>
    </div>
  );
}

function renderPartnerSlot({href, tracking_label, title, image}) {
  return (
    <div className="wds-global-navigation__partner-slot">
      <a
        className="wds-global-navigation__partner-slot-link"
        href={href}
        data-tracking-label={tracking_label}
      >
        <img
          className="wds-global-navigation__partner-slot-image"
          src={image}
          alt={title}
        />
      </a>
    </div>
  );
}

function renderLogo({href, tracking_label, title, edition}) {
  return (
    <a
      href={href}
      className="wds-global-navigation__logo"
      data-tracking-label={tracking_label}
      title={title}
    >
      <svg className="wds-global-navigation__logo-image wds-is-wds-company-logo-fandom-white">
        <use xlinkHref="#wds-company-logo-fandom-white" />
      </svg>
      {edition && (
        <span className="wds-global-navigation__edition-text">{edition}</span>
      )}
    </a>
  );
}

function renderLogout({href, tracking_label, title}) {
  return (
    <div className="">
      <div id="global-navigation-user-sign-out">
        <form method="POST" action={href}>
          <button
            className="wds-sign-out__button"
            data-tracking-label={tracking_label}
            type="submit"
          >
            {title}
          </button>
        </form>
      </div>
    </div>
  );
}

function renderFullButton({href, tracking_label, title, caption}, isSecondary) {
  return (
    <React.Fragment>
      {caption && (
        <div className="wds-global-navigation__user-menu-dropdown-caption">
          {caption}
        </div>
      )}
      <a
        className={`wds-is-full-width wds-button ${isSecondary && 'wds-is-secondary'}`}
        href={href}
        data-tracking-label={tracking_label}
      >
        {title}
      </a>
    </React.Fragment>
  );
}

/* eslint-enable react/prop-types */

const Link = ({link, ...props}) => {
  if (!link) {
    return null;
  }

  const {type} = link;

  switch (type) {
    case 'link-nav':
      return renderLinkNav(link, props);
    case 'link-text':
      return renderLinkText(link, props);
    case 'link-button':
      return renderLinkButton(link, props);
    case 'link-group':
      return renderLinkGroup(link);
    case 'link-partner-slot':
      return renderPartnerSlot(link);
    case 'link-logo':
      return renderLogo(link);
    case 'link-logout':
      return renderLogout(link);
    case 'link-full-button':
      return renderFullButton(link, false);
    case 'link-full-button-secondary':
      return renderFullButton(link, true);
    default:
    // eslint-disable-next-line no-console
      console.error('Unknown Link type: ', type, link);
  }
  return null;
};

Link.propTypes = {
  link: linkType,
};

Link.defaultProps = {
  link: null,
};

export default Link;
