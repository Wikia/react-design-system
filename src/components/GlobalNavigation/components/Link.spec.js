import React from 'react';
import renderer from 'react-test-renderer';
import {shallow} from 'enzyme';
import sinon from 'sinon';

import Link from './Link';

const emptyLink = {
  type: '',
  title: 'title',
  tracking_label: 'tracking_label',
};

/* eslint-disable no-alert */

test('Link should\'t render empty', () => {
  const component = renderer.create(
    <Link link={null} />
  );
  expect(component.toJSON()).toBe(null);
});

test('Link should\'t render unknown type', () => {
  // mock console.error
  const consoleSpy = jest.spyOn(console, 'error').mockImplementation(() => {});

  const unknownLink = {...emptyLink, type: '?'};
  const component = renderer.create(
    <Link link={unknownLink} />
  );
  expect(component.toJSON()).toBe(null);
  // check console.error
  expect(console.error).toHaveBeenCalled(); // eslint-disable-line no-console
  expect(consoleSpy.mock.calls[0][0]).toContain('Unknown Link type:');

  // restore global console
  consoleSpy.mockRestore();
});

test('Link should render type link-nav', () => {
  const link = {
    ...emptyLink,
    type: 'link-nav',
    href: 'http://example.com',
  };
  const component = renderer.create(
    <Link link={link} />
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('Link should render type link-text', () => {
  const link = {
    ...emptyLink,
    type: 'link-text',
    href: 'http://example.com',
  };
  const component = renderer.create(
    <Link link={link} />
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('Link should render type link-button', () => {
  const link = {
    ...emptyLink,
    type: 'link-button',
    href: 'http://example.com',
  };
  const component = renderer.create(
    <Link link={link} />
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('Link should render type link-group', () => {
  const link = {
    ...emptyLink,
    type: 'link-group',
    title: 'group',
    items: [
      {...emptyLink, type: 'link-text', href: 'http://example.com', tracking_label: '1'},
      {...emptyLink, type: 'link-button', href: 'http://example.com', tracking_label: '2'},
    ],
  };
  const component = renderer.create(
    <Link link={link} />
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('Link should render type link-partner-slot', () => {
  const link = {
    ...emptyLink,
    type: 'link-partner-slot',
    href: 'http://partner.com',
    image: 'http://partner.com/image.jpg',
  };
  const component = renderer.create(
    <Link link={link} />
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('Link should render type link-logo without edition', () => {
  const link = {
    ...emptyLink,
    type: 'link-logo',
    href: 'http://example.com',
  };
  const component = renderer.create(
    <Link link={link} />
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('Link should render type link-logo with edition', () => {
  const link = {
    ...emptyLink,
    type: 'link-logo',
    href: 'http://example.com',
    edition: 'UK',
  };
  const component = renderer.create(
    <Link link={link} />
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('Link should render type link-logout', () => {
  const link = {
    ...emptyLink,
    type: 'link-logout',
    href: 'http://example.com',
  };
  const component = renderer.create(
    <Link link={link} />
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('Link should render type link-full-button', () => {
  const link = {
    ...emptyLink,
    type: 'link-full-button',
    href: 'http://example.com',
  };
  const component = renderer.create(
    <Link link={link} />
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('Link should render type link-full-button-secondary with caption', () => {
  const link = {
    ...emptyLink,
    type: 'link-full-button-secondary',
    href: 'http://example.com',
    caption: 'A caption',
  };
  const component = renderer.create(
    <Link link={link} />
  );
  expect(component.toJSON()).toMatchSnapshot();
});
