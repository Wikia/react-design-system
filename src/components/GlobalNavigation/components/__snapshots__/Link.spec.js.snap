// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link should render type link-button 1`] = `
<a
  className="wds-global-navigation__link-button wds-button wds-is-secondary"
  data-tracking-label="tracking_label"
  href="http://example.com"
>
  title
</a>
`;

exports[`Link should render type link-full-button 1`] = `
<a
  className="wds-is-full-width wds-button false"
  data-tracking-label="tracking_label"
  href="http://example.com"
>
  title
</a>
`;

exports[`Link should render type link-full-button-secondary with caption 1`] = `
Array [
  <div
    className="wds-global-navigation__user-menu-dropdown-caption"
  >
    A caption
  </div>,
  <a
    className="wds-is-full-width wds-button wds-is-secondary"
    data-tracking-label="tracking_label"
    href="http://example.com"
  >
    title
  </a>,
]
`;

exports[`Link should render type link-group 1`] = `
<div
  className="wds-global-navigation__link-group wds-has-shadow wds-dropdown"
>
  <div
    className="wds-global-navigation__dropdown-toggle wds-global-navigation__link wds-dropdown__toggle"
  >
    <span>
      group
    </span>
    <svg
      className="wds-icon wds-icon-tiny wds-dropdown__toggle-chevron"
    >
      <use
        xlinkHref="#wds-icons-dropdown-tiny"
      />
    </svg>
  </div>
  <div
    className="wds-global-navigation__dropdown-content wds-dropdown__content"
  >
    <ul
      className="wds-list wds-is-linked"
    >
      <li>
        <a
          data-tracking-label="1"
          href="http://example.com"
        >
          title
        </a>
      </li>
      <li>
        <a
          className="wds-global-navigation__link-button wds-button wds-is-secondary"
          data-tracking-label="2"
          href="http://example.com"
        >
          title
        </a>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Link should render type link-logo with edition 1`] = `
<a
  className="wds-global-navigation__logo"
  data-tracking-label="tracking_label"
  href="http://example.com"
  title="title"
>
  <svg
    className="wds-global-navigation__logo-image wds-is-wds-company-logo-fandom-white"
  >
    <use
      xlinkHref="#wds-company-logo-fandom-white"
    />
  </svg>
  <span
    className="wds-global-navigation__edition-text"
  >
    UK
  </span>
</a>
`;

exports[`Link should render type link-logo without edition 1`] = `
<a
  className="wds-global-navigation__logo"
  data-tracking-label="tracking_label"
  href="http://example.com"
  title="title"
>
  <svg
    className="wds-global-navigation__logo-image wds-is-wds-company-logo-fandom-white"
  >
    <use
      xlinkHref="#wds-company-logo-fandom-white"
    />
  </svg>
</a>
`;

exports[`Link should render type link-logout 1`] = `
<div
  className=""
>
  <div
    id="global-navigation-user-sign-out"
  >
    <form
      action="http://example.com"
      method="POST"
    >
      <button
        className="wds-sign-out__button"
        data-tracking-label="tracking_label"
        type="submit"
      >
        title
      </button>
    </form>
  </div>
</div>
`;

exports[`Link should render type link-nav 1`] = `
<a
  className="wds-global-navigation__link"
  data-tracking-label="tracking_label"
  href="http://example.com"
>
  title
</a>
`;

exports[`Link should render type link-partner-slot 1`] = `
<div
  className="wds-global-navigation__partner-slot"
>
  <a
    className="wds-global-navigation__partner-slot-link"
    data-tracking-label="tracking_label"
    href="http://partner.com"
  >
    <img
      alt="title"
      className="wds-global-navigation__partner-slot-image"
      src="http://partner.com/image.jpg"
    />
  </a>
</div>
`;

exports[`Link should render type link-text 1`] = `
<a
  data-tracking-label="tracking_label"
  href="http://example.com"
>
  title
</a>
`;
